File: stop/vm.go
package stop

import (
	"fmt"

	"github.com/vcokltfre/stop/stop/instructions"
)

const (
	STACK_SIZE      = 1024
	CALL_STACK_SIZE = 64
	DEBUG           = false
)

type VM struct {
	stack        []int64
	stackTop     int
	callStack    []int
	callStackTop int
	program      []byte
	registers    []int64

	jumps map[uint16]int
	index int
}

func (v *VM) debug(data ...any) {
	if DEBUG {
		fmt.Println(append([]any{"[DEBUG]", fmt.Sprintf("%x", v.index)}, data...)...)
	}
}

func (v *VM) buildJumps() {
	index := 0

	for index < len(v.program) {
		curr := v.program[index]
		switch curr {
		case instructions.IHeaderHlt:
			index += instructions.ISizeHlt
		case instructions.IHeaderDbg:
			index += instructions.ISizeDbg
		case instructions.IHeaderMovLiteral:
			index += instructions.ISizeMovLiteral
		case instructions.IHeaderMovRegister:
			index += instructions.ISizeMovRegister
		case instructions.IHeaderPush:
			index += instructions.ISizePush
		case instructions.IHeaderDup:
			index += instructions.ISizeDup
		case instructions.IHeaderDrop:
			index += instructions.ISizeDrop
		case instructions.IHeaderSwap:
			index += instructions.ISizeSwap
		case instructions.IHeaderLd:
			index += instructions.ISizeLd
		case instructions.IHeaderSt:
			index += instructions.ISizeSt
		case instructions.IHeaderAdd:
			index += instructions.ISizeAdd
		case instructions.IHeaderSub:
			index += instructions.ISizeSub
		case instructions.IHeaderMul:
			index += instructions.ISizeMul
		case instructions.IHeaderDiv:
			index += instructions.ISizeDiv
		case instructions.IHeaderMod:
			index += instructions.ISizeMod
		case instructions.IHeaderLabel:
			index += instructions.ISizeLabel
			v.jumps[uint16(v.program[index-1])<<8|uint16(v.program[index-2])] = index
		case instructions.IHeaderCall:
			index += instructions.ISizeCall
		case instructions.IHeaderJmp:
			index += instructions.ISizeJmp
		case instructions.IHeaderJmpZ:
			index += instructions.ISizeJmp
		case instructions.IHeaderJmpNZ:
			index += instructions.ISizeJmp
		case instructions.IHeaderJmpP:
			index += instructions.ISizeJmp
		case instructions.IHeaderJmpN:
			index += instructions.ISizeJmp
		case instructions.IHeaderRet:
			index += instructions.ISizeRet
		case instructions.IHeaderPutN:
			index += instructions.ISizePutN
		case instructions.IHeaderPutC:
			index += instructions.ISizePutC
		default:
			panic("invalid instruction: " + fmt.Sprintf("%x", curr))
		}
	}
}

func (v *VM) stackPush(val int64) {
	if v.stackTop >= STACK_SIZE-1 {
		panic("stack overflow")
	}

	v.debug("pushing", v.stackTop, val)

	v.stackTop++
	v.stack[v.stackTop] = val
}

func (v *VM) stackPop() int64 {
	if v.stackTop < 0 {
		panic("stack underflow")
	}

	v.debug("popping", v.stackTop, v.stack[v.stackTop])

	v.stackTop--
	return v.stack[v.stackTop+1]
}

func (v *VM) callStackPush(val int) {
	if v.callStackTop >= CALL_STACK_SIZE-1 {
		panic("call stack overflow")
	}

	v.callStackTop++
	v.callStack[v.callStackTop] = val
}

func (v *VM) callStackPop() int {
	if v.callStackTop < 0 {
		panic("call stack underflow")
	}

	v.callStackTop--
	return v.callStack[v.callStackTop+1]
}

func (v *VM) getReg() uint8 {
	return v.program[v.index]
}

func (v *VM) getU16() uint16 {
	return uint16(v.program[v.index+1])<<8 | uint16(v.program[v.index])
}

func (v *VM) getI64() int64 {
	var i int64
	for j := 0; j < 8; j++ {
		i |= int64(v.program[v.index+j]) << uint64(j*8)
	}
	return i
}

func (v *VM) instDebug() {
	fmt.Println("TODO: implement debug instruction")
}

func (v *VM) instMovLiteral() {
	v.index += 1
	reg := v.getReg()
	v.index += 1
	val := v.getI64()
	v.registers[int(reg)] = val
	v.index += 8
}

func (v *VM) instMovRegister() {
	v.index += 1
	reg := v.getReg()
	v.index += 1
	src := v.getReg()
	v.registers[int(reg)] = v.registers[int(src)]
	v.index += 1
}

func (v *VM) instPush() {
	v.index++
	v.stackPush(v.getI64())
	v.index += 8
}

func (v *VM) instDup() {
	val := v.stackPop()
	v.stackPush(val)
	v.stackPush(val)
	v.index += instructions.ISizeDup
}

func (v *VM) instDrop() {
	v.stackPop()
	v.index += instructions.ISizeDrop
}

func (v *VM) instSwap() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a)
	v.stackPush(b)
	v.index += instructions.ISizeSwap
}

func (v *VM) instLd() {
	v.index += 1
	reg := v.getReg()
	v.stackPush(v.registers[int(reg)])
	v.index += 1
}

func (v *VM) instSt() {
	v.index += 1
	reg := v.getReg()
	v.registers[int(reg)] = v.stackPop()
	v.index += 1
}

func (v *VM) instAdd() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a + b)
	v.index += instructions.ISizeAdd
}

func (v *VM) instSub() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a - b)
	v.index += instructions.ISizeSub
}

func (v *VM) instMul() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a * b)
	v.index += instructions.ISizeMul
}

func (v *VM) instDiv() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a / b)
	v.index += instructions.ISizeDiv
}

func (v *VM) instMod() {
	a := v.stackPop()
	b := v.stackPop()
	v.stackPush(a % b)
	v.index += instructions.ISizeMod
}

func (v *VM) instJmp() {
	v.index += 1
	addr := v.getU16()
	v.index = v.jumps[addr]
}

func (v *VM) instJmpZ() {
	v.index += 1
	addr := v.getU16()
	val := v.stackPop()
	if val == 0 {
		v.index = v.jumps[addr]
	} else {
		v.index += 2
	}
}

func (v *VM) instJmpNZ() {
	v.index += 1
	addr := v.getU16()
	val := v.stackPop()
	if val != 0 {
		v.index = v.jumps[addr]
	} else {
		v.index += 2
	}
}

func (v *VM) instJmpP() {
	v.index += 1
	addr := v.getU16()
	val := v.stackPop()
	if val > 0 {
		v.index = v.jumps[addr]
	} else {
		v.index += 2
	}
}

func (v *VM) instJmpN() {
	v.index += 1
	addr := v.getU16()
	val := v.stackPop()
	if val < 0 {
		v.index = v.jumps[addr]
	} else {
		v.index += 2
	}
}

func (v *VM) instRet() {
	v.index = v.callStackPop()
}

func (v *VM) instPutN() {
	v.index += 1
	val := v.stackPop()
	fmt.Println(val)
}

func (v *VM) instPutC() {
	v.index += 1
	val := v.stackPop()
	fmt.Printf("%c", val)
}

func (v *VM) step() bool {
	if v.index >= len(v.program) {
		return true
	}

	curr := v.program[v.index]
	switch curr {
	case instructions.IHeaderHlt:
		return true
	case instructions.IHeaderDbg:
		v.instDebug()
		v.index += instructions.ISizeDbg
	case instructions.IHeaderMovLiteral:
		v.debug("mov literal")
		v.instMovLiteral()
	case instructions.IHeaderMovRegister:
		v.debug("mov register")
		v.instMovRegister()
	case instructions.IHeaderPush:
		v.debug("push")
		v.instPush()
	case instructions.IHeaderDup:
		v.debug("dup")
		v.instDup()
	case instructions.IHeaderDrop:
		v.debug("drop")
		v.instDrop()
	case instructions.IHeaderSwap:
		v.debug("swap")
		v.instSwap()
	case instructions.IHeaderLd:
		v.debug("ld")
		v.instLd()
	case instructions.IHeaderSt:
		v.debug("st")
		v.instSt()
	case instructions.IHeaderAdd:
		v.debug("add")
		v.instAdd()
	case instructions.IHeaderSub:
		v.debug("sub")
		v.instSub()
	case instructions.IHeaderMul:
		v.debug("mul")
		v.instMul()
	case instructions.IHeaderDiv:
		v.debug("div")
		v.instDiv()
	case instructions.IHeaderMod:
		v.debug("mod")
		v.instMod()
	case instructions.IHeaderLabel:
		v.debug("label")
		v.index += instructions.ISizeLabel
	case instructions.IHeaderCall:
		v.debug("call")
		v.index += 1
		v.callStackPush(v.index + 2)
		v.index = v.jumps[v.getU16()]
	case instructions.IHeaderJmp:
		v.debug("jmp")
		v.instJmp()
	case instructions.IHeaderJmpZ:
		v.debug("jmpz")
		v.instJmpZ()
	case instructions.IHeaderJmpNZ:
		v.debug("jmpnz")
		v.instJmpNZ()
	case instructions.IHeaderJmpP:
		v.debug("jmpp")
		v.instJmpP()
	case instructions.IHeaderJmpN:
		v.debug("jmpn")
		v.instJmpN()
	case instructions.IHeaderRet:
		v.debug("ret")
		v.instRet()
	case instructions.IHeaderPutN:
		v.debug("putn")
		v.instPutN()
	case instructions.IHeaderPutC:
		v.debug("putc")
		v.instPutC()
	default:
		panic("invalid instruction: " + fmt.Sprintf("%x", curr) + " at " + fmt.Sprintf("%x (%d)", v.index, v.index))
	}
	return false
}

func (v *VM) Run(code []byte) {
	v.stack = make([]int64, STACK_SIZE)
	v.callStack = make([]int, CALL_STACK_SIZE)
	v.program = code
	v.jumps = make(map[uint16]int)
	v.registers = make([]int64, 16)

	v.buildJumps()

	for {
		stop := v.step()
		if stop {
			break
		}
	}
}

File: csvm/vm.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#define STACK_SIZE 256
#define CALL_STACK_SIZE 256

// #define DEBUG

#ifdef DEBUG
#define debug(...) printf(__VA_ARGS__)
#else
#define debug(...)
#endif

#define IHeaderHlt 0x00         // Halt
#define IHeaderDbg 0x01         // Debug
#define IHeaderMovLiteral 0x08  // Move value
#define IHeaderMovRegister 0x09 // Move register
#define IHeaderPush 0x10        // Push value
#define IHeaderDup 0x11         // Duplicate value
#define IHeaderDrop 0x12        // Drop value
#define IHeaderSwap 0x13        // Swap values
#define IHeaderLd 0x20          // Load value
#define IHeaderSt 0x21          // Store value
#define IHeaderAdd 0x30         // Add
#define IHeaderSub 0x31         // Subtract
#define IHeaderMul 0x32         // Multiply
#define IHeaderDiv 0x33         // Divide
#define IHeaderMod 0x34         // Modulo
#define IHeaderLabel 0xA0       // Label
#define IHeaderCall 0xA1        // Call
#define IHeaderJmp 0xA2         // Jump
#define IHeaderJmpZ 0xA3        // Jump if zero
#define IHeaderJmpNZ 0xA4       // Jump if not zero
#define IHeaderJmpP 0xA5        // Jump if positive
#define IHeaderJmpN 0xA6        // Jump if negative
#define IHeaderRet 0xA7         // Return
#define IHeaderPutN 0xB0        // Put number
#define IHeaderPutC 0xB1        // Put character

const unsigned int ISizeHlt = 1;         // {header}
const unsigned int ISizeDbg = 1;         // {header}
const unsigned int ISizeMovLiteral = 10; // {header, reg, value[8]}
const unsigned int ISizeMovRegister = 3; // {header, reg, source}
const unsigned int ISizePush = 9;        // {header, value[8]}
const unsigned int ISizeDup = 1;         // {header}
const unsigned int ISizeDrop = 1;        // {header}
const unsigned int ISizeSwap = 1;        // {header}
const unsigned int ISizeLd = 2;          // {header, reg}
const unsigned int ISizeSt = 2;          // {header, reg}
const unsigned int ISizeAdd = 1;         // {header}
const unsigned int ISizeSub = 1;         // {header}
const unsigned int ISizeMul = 1;         // {header}
const unsigned int ISizeDiv = 1;         // {header}
const unsigned int ISizeMod = 1;         // {header}
const unsigned int ISizeLabel = 3;       // {header, label[2]}
const unsigned int ISizeCall = 3;        // {header, label[2]}
const unsigned int ISizeJmp = 3;         // {header, label[2]}
const unsigned int ISizeJmpZ = 3;        // {header, label[2]}
const unsigned int ISizeJmpNZ = 3;       // {header, label[2]}
const unsigned int ISizeJmpP = 3;        // {header, label[2]}
const unsigned int ISizeJmpN = 3;        // {header, label[2]}
const unsigned int ISizeRet = 1;         // {header}
const unsigned int ISizePutN = 1;        // {header}
const unsigned int ISizePutC = 1;        // {header}

int64_t stack[STACK_SIZE];
uint64_t sp = 0;

uint64_t call_stack[CALL_STACK_SIZE];
uint64_t csp = 0;

uint64_t jumps[1 << 16];
uint64_t registers[16];

uint64_t ip = 0;

void build_jumps(uint8_t *buffer, long size)
{
    uint64_t ip = 0;

    while (ip < size)
    {
        switch (buffer[ip])
        {
        case IHeaderHlt:
            ip += ISizeHlt;
            break;
        case IHeaderDbg:
            ip += ISizeDbg;
            break;
        case IHeaderMovLiteral:
            ip += ISizeMovLiteral;
            break;
        case IHeaderMovRegister:
            ip += ISizeMovRegister;
            break;
        case IHeaderPush:
            ip += ISizePush;
            break;
        case IHeaderDup:
            ip += ISizeDup;
            break;
        case IHeaderDrop:
            ip += ISizeDrop;
            break;
        case IHeaderSwap:
            ip += ISizeSwap;
            break;
        case IHeaderLd:
            ip += ISizeLd;
            break;
        case IHeaderSt:
            ip += ISizeSt;
            break;
        case IHeaderAdd:
            ip += ISizeAdd;
            break;
        case IHeaderSub:
            ip += ISizeSub;
            break;
        case IHeaderMul:
            ip += ISizeMul;
            break;
        case IHeaderDiv:
            ip += ISizeDiv;
            break;
        case IHeaderMod:
            ip += ISizeMod;
            break;
        case IHeaderLabel:
            ip += ISizeLabel;
            jumps[(buffer[ip - 1] << 8) | buffer[ip - 2]] = ip;
            break;
        case IHeaderCall:
            ip += ISizeCall;
            break;
        case IHeaderJmp:
            ip += ISizeJmp;
            break;
        case IHeaderJmpZ:
            ip += ISizeJmpZ;
            break;
        case IHeaderJmpNZ:
            ip += ISizeJmpNZ;
            break;
        case IHeaderJmpP:
            ip += ISizeJmpP;
            break;
        case IHeaderJmpN:
            ip += ISizeJmpN;
            break;
        case IHeaderRet:
            ip += ISizeRet;
            break;
        case IHeaderPutN:
            ip += ISizePutN;
            break;
        case IHeaderPutC:
            ip += ISizePutC;
            break;
        default:
            printf("Error: unknown instruction %d\n", buffer[ip]);
            return;
        }
    }

    ip = 0;
}

void push(uint64_t value)
{
    if (sp >= STACK_SIZE)
    {
        printf("Error: stack overflow\n");
        exit(1);
    }

    stack[sp++] = value;
}

uint64_t pop()
{
    if (sp <= 0)
    {
        printf("Error: stack underflow\n");
        exit(1);
    }

    return stack[--sp];
}

uint64_t peek()
{
    if (sp <= 0)
    {
        printf("Error: stack underflow\n");
        exit(1);
    }

    return stack[sp - 1];
}

void call(uint64_t address)
{
    if (csp >= CALL_STACK_SIZE)
    {
        printf("Error: call stack overflow\n");
        exit(1);
    }

    call_stack[csp++] = address;
}

uint64_t ret()
{
    if (csp <= 0)
    {
        printf("Error: call stack underflow\n");
        exit(1);
    }

    return call_stack[--csp];
}

uint16_t read_u16(uint8_t *buffer, uint64_t offset)
{
    return (buffer[offset + 1] << 8) | buffer[offset];
}

int64_t read_i64(uint8_t *buffer, uint64_t offset) {
    return (
        (uint64_t)buffer[offset + 0] << 0 |
        (uint64_t)buffer[offset + 1] << 8 |
        (uint64_t)buffer[offset + 2] << 16 |
        (uint64_t)buffer[offset + 3] << 24 |
        (uint64_t)buffer[offset + 4] << 32 |
        (uint64_t)buffer[offset + 5] << 40 |
        (uint64_t)buffer[offset + 6] << 48 |
        (uint64_t)buffer[offset + 7] << 56
    );
}

void i_mov_literal(uint8_t *buffer)
{
    ip++;
    uint8_t reg = buffer[ip];
    ip++;
    registers[reg] = read_i64(buffer, ip);
    ip += 8;
}

void i_mov_register(uint8_t *buffer)
{
    ip++;
    uint8_t reg = buffer[ip++];
    uint8_t source = buffer[ip++];
    registers[reg] = registers[source];
}

void i_push(uint8_t *buffer)
{
    ip++;
    push(read_i64(buffer, ip));
    ip += 8;
}

void i_dup(uint8_t *buffer)
{
    ip++;
    push(peek());
}

void i_drop(uint8_t *buffer)
{
    ip++;
    pop();
}

void i_swap(uint8_t *buffer)
{
    ip++;
    uint64_t a = pop();
    uint64_t b = pop();
    push(a);
    push(b);
}

void i_ld(uint8_t *buffer)
{
    ip++;
    uint8_t reg = buffer[ip++];
    push(registers[reg]);
}

void i_st(uint8_t *buffer)
{
    ip++;
    uint8_t reg = buffer[ip++];
    registers[reg] = pop();
}

void i_add(uint8_t *buffer)
{
    uint64_t a = pop();
    uint64_t b = pop();
    push(a + b);

    ip += ISizeAdd;
}

void i_sub(uint8_t *buffer)
{
    uint64_t a = pop();
    uint64_t b = pop();
    push(a - b);

    ip += ISizeSub;
}

void i_mul(uint8_t *buffer)
{
    uint64_t a = pop();
    uint64_t b = pop();
    push(a * b);

    ip += ISizeMul;
}

void i_div(uint8_t *buffer)
{
    uint64_t a = pop();
    uint64_t b = pop();
    push(a / b);

    ip += ISizeDiv;
}

void i_mod(uint8_t *buffer)
{
    uint64_t a = pop();
    uint64_t b = pop();
    push(a % b);

    ip += ISizeMod;
}

void i_jmp(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    ip = jumps[addr];
}

void i_jmpz(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    ip += 2;
    if (pop() == 0)
    {
        ip = jumps[addr];
    }
}

void i_jmpnz(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    ip += 2;
    if (pop() != 0)
    {
        ip = jumps[addr];
    }
}

void i_jmpp(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    ip += 2;
    if (pop() > 0)
    {
        ip = jumps[addr];
    }
}

void i_jmpn(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    ip += 2;
    if (pop() < 0)
    {
        ip = jumps[addr];
    }
}

void i_call(uint8_t *buffer)
{
    ip++;
    uint16_t addr = read_u16(buffer, ip);
    call(ip+2);
    ip = jumps[addr];
}

void i_ret(uint8_t *buffer)
{
    ip = ret();
}

void i_putn(uint8_t *buffer)
{
    ip++;
    printf("%ld\n", pop());
}

void i_putc(uint8_t *buffer)
{
    ip++;
    printf("%c", (char)pop());
}

int run(uint8_t *buffer, long size)
{
    build_jumps(buffer, size);

    while (ip < size) {
        switch (buffer[ip]) {
        case IHeaderHlt:
            return pop();
        case IHeaderDbg:
            printf("ip=%ld, sp=%ld, csp=%ld\n", ip, sp, csp);
            ip += ISizeDbg;
            break;
        case IHeaderMovLiteral:
            debug("movl %d %ld\n", buffer[ip + 1], read_i64(buffer, ip + 2));
            i_mov_literal(buffer);
            break;
        case IHeaderMovRegister:
            debug("movr %d %d\n", buffer[ip + 1], buffer[ip + 2]);
            i_mov_register(buffer);
            break;
        case IHeaderPush:
            debug("push %ld\n", read_i64(buffer, ip + 1));
            i_push(buffer);
            break;
        case IHeaderDup:
            debug("dup\n");
            i_dup(buffer);
            break;
        case IHeaderDrop:
            debug("drop\n");
            i_drop(buffer);
            break;
        case IHeaderSwap:
            debug("swap\n");
            i_swap(buffer);
            break;
        case IHeaderLd:
            debug("ld %d\n", buffer[ip + 1]);
            i_ld(buffer);
            break;
        case IHeaderSt:
            debug("st %d\n", buffer[ip + 1]);
            i_st(buffer);
            break;
        case IHeaderAdd:
            debug("add\n");
            i_add(buffer);
            break;
        case IHeaderSub:
            debug("sub\n");
            i_sub(buffer);
            break;
        case IHeaderMul:
            debug("mul\n");
            i_mul(buffer);
            break;
        case IHeaderDiv:
            debug("div\n");
            i_div(buffer);
            break;
        case IHeaderMod:
            debug("mod\n");
            i_mod(buffer);
            break;
        case IHeaderLabel:
            debug("label %d\n", read_u16(buffer, ip + 1));
            ip += ISizeLabel;
            break;
        case IHeaderCall:
            debug("call %d\n", read_u16(buffer, ip + 1));
            i_call(buffer);
            break;
        case IHeaderJmp:
            debug("jmp %d\n", read_u16(buffer, ip + 1));
            i_jmp(buffer);
            break;
        case IHeaderJmpZ:
            debug("jmpz %d\n", read_u16(buffer, ip + 1));
            i_jmpz(buffer);
            break;
        case IHeaderJmpNZ:
            debug("jmpnz %d\n", read_u16(buffer, ip + 1));
            i_jmpnz(buffer);
            break;
        case IHeaderJmpP:
            debug("jmpp %d\n", read_u16(buffer, ip + 1));
            i_jmpp(buffer);
            break;
        case IHeaderJmpN:
            debug("jmpn %d\n", read_u16(buffer, ip + 1));
            i_jmpn(buffer);
            break;
        case IHeaderRet:
            debug("ret\n");
            i_ret(buffer);
            break;
        case IHeaderPutN:
            debug("putn\n");
            i_putn(buffer);
            break;
        case IHeaderPutC:
            debug("putc\n");
            i_putc(buffer);
            break;
        default:
            printf("Error: unknown instruction %d\n", buffer[ip]);
            return 1;
        }
    }

    return 0;
}

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        printf("Usage: %s <file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp)
    {
        printf("Error: could not open file %s\n", argv[1]);
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    long size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    char *buffer = malloc(size);
    fread(buffer, size, 1, fp);
    fclose(fp);

    return run(buffer, size);
}
